### ISSUE TITLE
Jackson's @JsonIgnoreProperties annotation not inherited

### ISSUE THREAD

**Issue by runenielsen** [micronaut-projects/micronaut-serialization#550]
### Expected Behavior

When serializing a model bean with inheritance and a discriminator property marked to be ignored with `@JsonIgnoreProperties`, I expect the discriminator property to only exist once in the JSON output by the serialization.

It should work so that any value set in the discriminator property in the bean is ignored, and the discriminator property is instead set in the serialized output based on the class of the bean. This is how it works in Jackson Databind.

### Actual Behaviour

The discriminator property exists twice in the JSON output by the serialization.

One value is generated from the value is set in discriminator property in the bean and the other value is generated from the bean's class.

### Steps To Reproduce

1. Clone the repository https://github.com/runenielsen/jsonignorepropertiesnotworking
2. Run the test JsonIgnorePropertiesWithDiscriminatorTest and see that Micronaut Serialization behaves differently than Jackson Databind.
3. I have included the test JsonIgnorePropertiesSimpleTest with a simple test of `@JsonIgnoreProperties` to verify, that it works correctly for a simple model without inheritance.
4. You can also see the incorrect output, if you run the application and open the URL http://localhost:8080/test in a browser.
5. This seems to work correctly when using `@JsonIgnore` instead of `@JsonIgnoreProperties` by the way, but the model classes in this reproduction project are a stripped down version of models generated by micronaut-openapi, so this is not something easily changeable in the class. And Micronaut Serialization should work correctly with both these annotations.

### Environment Information

MacOS
JDK 17

### Example Application

https://github.com/runenielsen/jsonignorepropertiesnotworking

### Version

4.0.3

**Comment by yawkat** [comment]
this is actually unrelated to subtyping, the JsonIgnoreProperties is just not inherited from the superclass. Adding the annotation to the subclass as well works around this issue.

**Comment by runenielsen** [comment]
Thanks, @yawkat 

The classes with the problem in our case is generated by Micronaut's OpenAPI generator, so it's unfortunately not easy for us to use your workaround without having to do a hack in Gradle to replace parts of them after generation.

**Comment by graemerocher** [comment]
probably the fix has to go into the openapi generator to be honest

**Comment by runenielsen** [comment]
@graemerocher : I know Jackson is not a standard, but I think it would be beneficial, if Micronaut Serialization works as closely as possible to Jackson, just without reflection.

This would make switching from Jackson to Micronaut Serialization the easiest for people using Jackson... It also means you (the Micronaut team) have to do less code in other places to compensate for differences between the two.

Based on this I personally think it would be better to fix this issue in Micronaut Serialization, so the annotation works the same way as in Jackson - but that is just my two cents 🙂

**Comment by yawkat** [comment]
given that it's a jackson annotation i agree it should work like jackson

**Comment by graemerocher** [comment]
in that case the annotation mappers should add the `@Inherited` annotation as a stereotype since Jackson doesn't declare it

**Comment by graemerocher** [comment]
might also be worth reviewing which other Jackson annotations are automatically inherited even if they don't declare `@Inherited`

**Comment by dstepanov** [comment]
This case is unique because the ignored property is part of the superclass, so we should copy the ignore definition to the subtypes.

**Comment by dstepanov** [comment]
@runenielsen Yes, it should work the same.

I'm also surprised that the type discriminator value is set, according to the `JsonTypeInfo`:

```
    /**
     * Property that defines whether type identifier value will be passed
     * as part of JSON stream to deserializer (true), or handled and
     * removed by <code>TypeDeserializer</code> (false).
     * Property has no effect on serialization.
     *<p>
     * Default value is false, meaning that Jackson handles and removes
     * the type identifier from JSON content that is passed to
     * <code>JsonDeserializer</code>.
     *
     * @since 2.0
     */
    public boolean visible() default false;
```

The default value is false and should not be passed to the bean deserialization.

**Comment by dstepanov** [comment]
It looks like it is generated correctly for the Jackson project:
```java
@Schema(name = "BookInfo", description = "Object containg all the info about a book")
@JsonPropertyOrder({
    BookInfo.JSON_PROPERTY_NAME,
    BookInfo.JSON_PROPERTY_TYPE
})
@Generated(value = "io.micronaut.openapi.generator.JavaMicronautServerCodegen")
@JsonIgnoreProperties(
  value = "type", // ignore manually set type, it will be automatically generated by Jackson during serialization
  allowSetters = true // allows the type to be set during deserialization
)
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type", visible = true)
@JsonSubTypes({
    @JsonSubTypes.Type(value = BasicBookInfo.class, name = "BASIC"),
    @JsonSubTypes.Type(value = DetailedBookInfo.class, name = "DETAILED")
})
@Introspected
public class BookInfo {
```

@runenielsen Can you please report the bug in the generator?

**Comment by runenielsen** [comment]
@dstepanov : Can you explain which bug it is you want me to report in the generator?

And if you mean the bug I reported in this issue, I don't see how it can be a bug in the generator - my example project doesn't use the generator...

**Comment by dstepanov** [comment]
Your project that generates the beans doesn't generate `visible = true` for Micronaut Serialization (but the Jackson one has it), not sure why is it happening. This means that `BookInfo` would not receive the descriminator property.

**Comment by runenielsen** [comment]
@dstepanov : This is the example project, I have given for this issue: https://github.com/runenielsen/jsonignorepropertiesnotworking

As far as I can tell, it doesn't use Micronaut's OpenAPI generator at all, and it uses the same models for Jackson and Micronaut Serialization. And it works for Jackson and not for Micronaut Serialization...

Which project are you looking at?

**Comment by dstepanov** [comment]
I'm looking at your other issue https://github.com/micronaut-projects/micronaut-serialization/issues/581
I have noticed that for the Micronaut Serialization it generates:
```java
@Serdeable
@JsonPropertyOrder({
    BookInfo.JSON_PROPERTY_NAME,
    BookInfo.JSON_PROPERTY_TYPE
})
@Generated(value = "io.micronaut.openapi.generator.JavaMicronautClientCodegen")
@JsonIgnoreProperties(
  value = "type", // ignore manually set type, it will be automatically generated by Jackson during serialization
  allowSetters = true // allows the type to be set during deserialization
)
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
@JsonSubTypes({
    @JsonSubTypes.Type(value = BasicBookInfo.class, name = "BASIC"),
    @JsonSubTypes.Type(value = DetailedBookInfo.class, name = "DETAILED")
})
@Introspected
public class BookInfo {
```

Notice how `visible = true` is missing compared to the output I posted above, generated by your Jackson project. I'm not sure why is it missing, but it looks like a bug in some generator.

**Comment by runenielsen** [comment]
Alright, but if we focus on the project I have given as an example here, would you not agree, that this is a bug in Micronaut Serialization?

It can't be a bug in the OpenAPI generator, since the OpenAPI generator is not used at all in the project, I have given as an example (https://github.com/runenielsen/jsonignorepropertiesnotworking).

So if the other issue #581 is a bug in the OpenAPI generator, this can't be a duplicate of it.

**Comment by yawkat** [comment]
@runenielsen it's clearly a bug in micronaut-serialization too, but there can be two bugs.

**Comment by dstepanov** [comment]
Right now, we have a few issues:
- Ignore properties from the superclass annotated with `@JsonIgnoreProperties`
- Supporting `visible` on `@JsonTypeInfo`
- Supporting `allowGetters` and `allowSetters` on `@JsonIgnoreProperties`

After that is implemented, the model generated in the Jackson project should work, but the project attached here would not work because of the missing `visible = true` here:

https://github.com/runenielsen/jsonignorepropertiesnotworking/blob/main/src/main/java/example/micronaut/BookInfo.java#L11-L25

**Comment by dstepanov** [comment]
I don't think Jackson works without `visible = true`. Please try.

**Comment by runenielsen** [comment]
@dstepanov 

I have just tried (again), and Jackson works fine in my attached reproduction project.

You can do it yourself like this:

1.  Clone the repository https://github.com/runenielsen/jsonignorepropertiesnotworking
2. Run the test JsonIgnorePropertiesWithDiscriminatorTest and see that Micronaut Serialization behaves differently than Jackson Databind. There is one test case for Micronaut Serialization and one for Jackson - they produce different output on serialization.

See also "Steps To Reproduce" above for more detail.

**Comment by dstepanov** [comment]
You are missing the point, I'm talking about `visible = true`:

See example:
```java
    @Test
    void testJacksonSerializationWithJsonIgnoreProperties() throws IOException {
        // It is intentional for the purpose of the test, that I have set the discriminator property to something that
        // doesn't match the class of the bean here. The value shouldn't matter, since it should be ignored.
        var testModel = new BasicBookInfo("Michael Ende", "The Neverending Story", BookInfo.TypeEnum.DETAILED);

        com.fasterxml.jackson.databind.ObjectMapper objectMapper1 = new com.fasterxml.jackson.databind.ObjectMapper();
        var testModelStr = objectMapper1.writeValueAsString(testModel);

        Assertions.assertEquals("{\"type\":\"BASIC\",\"author\":\"Michael Ende\",\"name\":\"The Neverending Story\"}", testModelStr);

        BookInfo bookInfo = objectMapper1.readValue("{\"type\":\"BASIC\",\"author\":\"Michael Ende\",\"name\":\"The Neverending Story\"}", BookInfo.class);

        Assertions.assertEquals("BASIC", bookInfo.getType().name());
    }
```

Fails with:
```java
@Serdeable
@JsonPropertyOrder({
        "name",
        "type"
})
@JsonIgnoreProperties(
  value = "type",
  allowSetters = true
)
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type")
@JsonSubTypes({
        @JsonSubTypes.Type(value = DetailedBookInfo.class, name = "DETAILED"),
        @JsonSubTypes.Type(value = BasicBookInfo.class, name = "BASIC"),
})
@Introspected
public class BookInfo
```

Passes with:
```java
@Serdeable
@JsonPropertyOrder({
        "name",
        "type"
})
@JsonIgnoreProperties(
  value = "type",
  allowSetters = true
)
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type", visible = true)
@JsonSubTypes({
        @JsonSubTypes.Type(value = DetailedBookInfo.class, name = "DETAILED"),
        @JsonSubTypes.Type(value = BasicBookInfo.class, name = "BASIC"),
})
@Introspected
public class BookInfo
```

**Comment by runenielsen** [comment]
@dstepanov : To be honest, I don't think you are getting my point... **_This issue is about serialization_**. You have changed the model and test case in my reproduction project and added some code that fails, when deserializing. I don't understand how this is relevant to this issue, which is about a bug when **_serializing_**.

I think your observations are relevant for the other issue #581 I reported, but these two issues are not duplicates. So in my opinion, all the comments you added to this issue today, should have been made on the other issue.

To sum it up for this issue: The Jackson test case in my unchanged reproduction project works and the identical Micronaut Serialization test case fails for the exact same model definitions. This must mean there is a difference in how Jackson and Micronaut serialization works. There shouldn't be when serializing models using Jackson annotations.

**Comment by dstepanov** [comment]
I just wanted to notify you about the other issue I found while investigating this one and the other you reported because otherwise, you are going to report that your use-case still doesn't work after I fixed them.
